src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs

```diff
@@ -37,23 +40,50 @@ static VisualStudioInstanceFactory()
                 throw new PlatformNotSupportedException("The Visual Studio Integration Test Framework is only supported on Visual Studio 15.0 and later.");
             }
 
-            // This looks like it is pointless (since we are returning an assembly that is already loaded) but it is actually required.
-            // The BinaryFormatter, when invoking 'HandleReturnMessage', will end up attempting to call 'BinaryAssemblyInfo.GetAssembly()',
-            // which will itself attempt to call 'Assembly.Load()' using the full name of the assembly for the type that is being deserialized.
-            // Depending on the manner in which the assembly was originally loaded, this may end up actually trying to load the assembly a second
-            // time and it can fail if the standard assembly resolution logic fails. This ensures that we 'succeed' this secondary load by returning
-            // the assembly that is already loaded.
-            AppDomain.CurrentDomain.AssemblyResolve += (sender, eventArgs) => {
-                Debug.WriteLine($"'{eventArgs.RequestingAssembly}' is attempting to resolve '{eventArgs.Name}'");
-                var resolvedAssembly = AppDomain.CurrentDomain.GetAssemblies().Where((assembly) => assembly.FullName.Equals(eventArgs.Name)).SingleOrDefault();
-
-                if (resolvedAssembly != null)
+            AppDomain.CurrentDomain.AssemblyResolve += AssemblyResolveHandler;
+            AppDomain.CurrentDomain.FirstChanceException += FirstChanceExceptionHandler;
+        }
+
+        private static void FirstChanceExceptionHandler(object sender, FirstChanceExceptionEventArgs eventArgs)
+        {
+            try
+            {
+                if (eventArgs.Exception is XunitException)
                 {
-                    Debug.WriteLine("The assembly was already loaded!");
+                    var assemblyPath = typeof(VisualStudioInstanceFactory).Assembly.Location;
+                    var assemblyDirectory = Path.GetDirectoryName(assemblyPath);
+                    var fileName = $"{eventArgs.Exception.GetType().Name}-{DateTime.Now:HH.mm.ss}.png";
```

Comment:
Ah, a custom [BeforeAfterTestAttribute](https://github.com/xunit/xunit/blob/master/src/xunit.core/Sdk/BeforeAfterTestAttribute.cs) might be just the thing.